docs/design/proposals/destructive-operations/destructive-operations.md

```diff
@@ -0,0 +1,202 @@
+
+This design document proposes different mechanisms to address each of the following concerns:
+
+* Protection against deletion
+* Control over AWS removal policy
+
+## Delete Protection
+
+### Requirements
+
+1. Multiple levels of granularity over resource protection (namespace-wide and resource-wide)
+2. Protection should be defined per-service (if namespace-wide)
+3. Controller must strictly check for existence of resource protection before any deletion action
+4. The controller must not edit any of the resource protection marks
+
+### Investigation
+
+Kubernetes finalizers are the standard way to protect a resource from deletion. A finalizer placed on a resource prohibits it from being deleted out of etcd until it is removed - either through code or by manually editing the resource. ACK already places finalizers on every resource, however these are used to guard the custom resource object being deleted before the the underlying AWS resource. Finalizers, however, do not apply in a cascading effect to resources below them ie. a finalizer placed on a namespace will not protect the deletion of a custom resource within it. This violates requirement 1, as there is only resource-wide granularity over resource protection. 
+
+Finalizers do not fit with requirement 3. If the user places a finalizer on an ACK resource, then issues a deletion of that resource, the ACK controller will immediately delete the AWS resource and then remove its own finalizer, regardless of any other finalizer on the object. The current ACK controller does not look for other finalizers on the object before running its delete logic. Only the K8s control plane uses finalizers to protect it from deletion in the key-value store, it does not limit the interaction of custom controllers.
+
+There is on ongoing issue regarding similar functionality in the [Kubernetes Github repository](https://github.com/kubernetes/kubernetes/issues/10179).
+There is a feature request for the same function in the [Azure service operator repository](https://github.com/Azure/azure-service-operator/issues/1633). 
+
+## AWS Removal Policy
+
+### Requirements
+
+5. Multiple levels of granularity over resource protection (controller-wide, namespace-wide and resource-wide)
+6. Protection should be defined per-service (if namespace-wide)
+7. Users are able to define either of:
+    - The AWS resource should be deleted along with the CR
+    - The AWS resource is left in-tact after the CR is deleted
+
+### Investigation
+
+A number of K8s resources already have similar functionality:
+
+* Helm uses an annotation: `helm.sh/resource-policy: keep`
+* Google Config Connector uses an annotation: `cnrm.cloud.google.com/deletion-policy: abandon`
+* Skaffold uses a label: `skaffold.dev/cleanup: "false"`
```

Comment:
There were also proposals originally to use an annotation/label for our resource adoption functionality. Instead of doing the annotation/label on a CR, we chose a user experience where the user needs to explicitly adopt an AWS resource by creating a `services.k8s.aws/AdoptedResource` CR. I continue to believe that we made the right decision in this regard: an explicit adoption action (via the `kubectl create -f <services.k8s.aws/AdoptedResource>` process) is the *least surprising, hardest to misuse interface*.

For the reverse of adoption action, I would prefer to have a similarly explicit action, essentially an "unadopt" or "unmanage" action. We could do this in a couple ways:

* Create a new kubectl plugin/verb called `kubectl unmanage` that the user could call to delete the CR but keep the AWS resource around
* `kubectl delete services.k8s.aws/AdoptedResource/<resource arn>` and have the AdoptedResourceReconciler run some "delete the CR but keep the AWS resource around" code